File: DCEPC/DCE204.java
import java.util.*;
import java.math.BigInteger;
import java.io.*;
class DCE204{
	static BufferedReader r;
	static BigInteger b;
	static BigInteger m=BigInteger.valueOf(1000000009);
	//static long val,a,b;
	//static BigInteger power(BigInteger a,BigInteger b){
	//    if (b==BigInteger.valueOf(1)) return a;
	//    if(b==BigInteger.valueOf(0)) return BigInteger.valueOf(1);
	//    BigInteger temp=power(a,b.divide(BigInteger.valueOf(2)));
	//    if(b.and(BigInteger.valueOf(1))==1)
	//    return temp.multiply(BigInteger.valueOf(temp));
	//    return temp.multiply(BigInteger.valueOf(temp));
	//}
	static BigInteger func(long n){
	     if(n==1||n==0||n==2)return BigInteger.valueOf(1);
	     if(n<5)return BigInteger.valueOf(3);
	     if(n<7)return BigInteger.valueOf(8);
	     if(n<=10)return BigInteger.valueOf(15);
	     else {
			b=BigInteger.valueOf(2*(int)ceil(n/4));
			b.multiply(BigInteger.valueOf(func((int)ceil(n/5))));
			b.multiply(BigInteger.valueOf(func((int)ceil(n/10))));
		}
	}


	public static void main(String args[]){
		int t=0;
    		long n=0;
    		r=new BufferedReader(new InputStreamReader(System.in));
		try{t=Integer.parseInt(r.readLine());}catch(Exception e){}
    		for(;t>0;t--){
                  try{n=Long.parseLong(r.readLine());}catch(Exception e){}
                  //val=func(n);
		  BigInteger c=BigInteger.valueOf(func(n));
		  BigInteger a=c.multiply(5);
                  BigInteger b=a.subtract(c);
                  //printf("%lld %lld",a,b);
                  BigInteger max=BigInteger.valueOf(0);
                  //if(power(a,b)>power(b,a)){max=power(a,b);}
                  //max=b.pow();
		a=a.mod(m);b=b.mod(m);
		max=b.pow(a.intValue());
                  max=max.mod(m);
			
                  System.out.println(max);
                 
    }
	}


}

File: DCEPC/DCE204.cpp
#include<stdio.h>
#include<cmath>
#include<iostream>
using namespace std;

//#define m 1000000009;
inline int gcd(long long int a,long long b){if(b==0)return a;else return gcd(b,a%b);}
long long int m=1000000000;
long long int a,b;
long long int power(long long int a,long long int b){
    if (b==1) return a;
    if(b==0) return 1;
    long long int temp=power(a,b/2);
    if(b&1)
    return (((temp*temp)%m)*a)%m;
    return (temp*temp)%m;
}
long long int func(long long int n){
     if(n==1||n==0||n==2)return 1;
     if(n<5)return 3;
     if(n<7)return 15;
     if(n<=10)return 105;
     //long long int q,w,e;
     //q=power(2,(long long int)ceil(n/4));
     //double r=n/5.0;
     //double t=n/10.0;
     //long long int z,x;
     //if(r>n/5)z=(long long int)r+1;
     //if(t>n/10)x=(long long int)t+1;
     //w=func(n/5);
     //e=func((long long int)ceil(n/10));
     //printf("%lld\t%f\t%lld\n",z,r,x);
     //return (((power(2,(long long int)ceil(n/4))*func((long long int)ceil(n/5)))%m)*func((long long int)ceil(n/10)))%m;
     else return(((((((power(2,(int)ceil(n/4))))*func((int)ceil(n/5))))*func((int)ceil(n/10)))));
     
}
long long int val;
int main(){
    int t;
    long long int n=0;
    scanf("%d",&t);
    for(;t>0;t--){
                  scanf("%lld",&n);
                  if(n>20){printf("%d\n",0);continue;}
                  val=func(n);
                  //printf("%lld\n",val); 
                  //a=5*val;
                  //b=a-val;
                  //printf("%lld %lld",a,b);
                  long long int tempi=0;
                  long long int max=0;
                  for(a=val;a<=5*val;a++){
                                         for(b=val;b<=5*val;b++){
                                                                 if(gcd(a,b)==val){
                                                                                   tempi=power(a,b);        
                                                                                   if(tempi>max)max=tempi;          
                                                                 }                        
                                         }                       
                  }
                  //else max=power(b,a);
                  //max=power(a,b);
                  printf("%lld\n",max);
                 
    }
    return 0;    
}
